cornac/models/causalrec/recom_causalrec.py

```diff
@@ -0,0 +1,400 @@
+# Copyright 2018 The Cornac Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ============================================================================
+
+import numpy as np
+from tqdm.auto import tqdm
+from scipy.special import expit
+
+from ..recommender import Recommender
+from ...exception import CornacException
+from ...exception import ScoreException
+from ...utils import fast_dot
+from ...utils.common import intersects
+from ...utils import get_rng
+from ...utils.init_utils import zeros, xavier_uniform
+
+
+class CausalRec(Recommender):
+    """CausalRec: Causal Inference for Visual Debiasing in Visually-Aware Recommendation
+
+    Parameters
+    ----------
+    k: int, optional, default: 10
+        The dimension of the gamma latent factors.
+
+    k2: int, optional, default: 10
+        The dimension of the theta latent factors.
+
+    n_epochs: int, optional, default: 20
+        Maximum number of epochs for SGD.
+
+    batch_size: int, optional, default: 100
+        The batch size for SGD.
+
+    learning_rate: float, optional, default: 0.001
+        The learning rate for SGD.
+
+    lambda_w: float, optional, default: 0.01
+        The regularization hyper-parameter for latent factor weights.
+
+    lambda_b: float, optional, default: 0.01
+        The regularization hyper-parameter for biases.
+
+    lambda_e: float, optional, default: 0.0
+        The regularization hyper-parameter for embedding matrix E and beta prime vector.
+
+    use_gpu: boolean, optional, default: True
+        Whether or not to use GPU to speed up training.
+
+    trainable: boolean, optional, default: True
+        When False, the model is not trained and Cornac assumes that the model already \
+        pre-trained (U and V are not None).
+
+    verbose: boolean, optional, default: True
+        When True, running logs are displayed.
+
+    init_params: dictionary, optional, default: None
+        Initial parameters, e.g., init_params = {'Bi': beta_item, 'Gu': gamma_user,
+        'Gi': gamma_item, 'Tu': theta_user, 'E': emb_matrix, 'Bp': beta_prime}
+
+    seed: int, optional, default: None
+        Random seed for weight initialization.
+
```

Comment:
Could you provide documentation for the two params `mean_feat` and `tanh`?

Reply:
Sure. Please see whether the second last commit is sufficient.